### ISSUE TITLE
GPL but no source?

### ISSUE THREAD

**Issue by h1z1** [vvoovv/blender-geo#89]
Curious, this project is referenced all over the place as the GPL addon to import data from Openstreetmap, yet I don't see any source available.  Appears commercial only?

**Comment by vvoovv** [comment]
It's totally ok to sell open source. Please read the article at gnu.org: [Selling Free Software](https://www.gnu.org/philosophy/selling.en.html)

Just a short quote from the article:

> Many people believe that the spirit of the GNU Project is that you should not charge money for distributing copies of software, or that you should charge as little as possible—just enough to cover the cost. This is a misunderstanding.

Selling GPL addon is only way for me to fund the development.

**Comment by h1z1** [comment]
Indeed.  It's not entirely clear what license this is released under?  Guess I'm more surprised such a project would exist on Github of all places.  The redistribution costs come from a time when logistical costs were more of an issue.  CDs and maintaining your own ftp site for example.

**Comment by vazhnov** [comment]
I love to support opensource project by money, even if I can install it for free (for example — OSMand, I bought 3 licenses: for me, for my wife and for my friend). But I don't see GPL sources here. Will I get them if I will pay here https://gumroad.com/l/blender-osm ?

**Comment by vvoovv** [comment]
Sure! You'll get all sources under GPL license!

**Comment by ghost** [comment]
Hi all,

Yes, the source code is included in the zip package that you get when you buy it.
Regards,

**Comment by JeremyBYU** [comment]
I agree that this is a good model to fund the development (I am a paying customer).  However I do have a question to make sure I understand the license correctly.

I received the source code and license from the purchased product (GPL v3).  From my understanding I am free to modify the source code and then **distribute** it also. In fact, if I distribute my modified plugin I **must** provide the source code (just as you @vvoovv does).  Is that correct? I don't think it should affect your revenue stream because yours is the officially one (with support).

The reason I am asking is because I have modified the code (for my research) to allow the user to specify the Spatial Reference System (basically the [coordinate system](http://pro.arcgis.com/en/pro-app/help/mapping/properties/coordinate-systems-and-projections.htm)) of the map projection. I believe right now its just using a Transverse Mercator, which causes distortion for the world objects (building shape, area, etc.) when they are far from the central meridian.  The plugin allows the user to choose whatever map projection they want.  Along this note, do you take pull requests @vvoovv? Maybe we could find a way to add something like this (the only problem is the reliance upon third party plugins - ones with binary dependencies!).  I also added the ability to add an automatic UV texture to buildings; when importing fbx models into [Unreal Engine](https://forums.unrealengine.com/development-discussion/content-creation/10685-imported-models-appearing-black-in-my-environment) the models were black because they lacked the [UV texture ](https://en.wikibooks.org/wiki/Blender_3D:_Noob_to_Pro/UV_Map_Basics).

So to recap 2 questions:

1. Was I right that I can publish the modified source code?
2. Do you take pull requests?

**Comment by vvoovv** [comment]
@JeremyBYU, yes you can publish the modified source code under the same license (GPL).
I'll contact you via e-mail concerning your modifications.

**Comment by JeremyBYU** [comment]
Great! Thanks for the quick response! I bought it under jdcasta@umich.edu.

**Comment by ghost** [comment]
Thank you all for the great work. I´m wondering whether you could share a
copy of your version, I´m quite interested in testing your spatial
reference system features. On a side note, which binary does your version
depend on? Is it necessary to test the coordinate system features?

Thank you

Kind regards,

Manuel

**Comment by vvoovv** [comment]
@JeremyBYU I'd suggest to release your custom map projection modification as an extension to the addon (i.e. an addon to the addon). In that way the extension will be always in sync with the official version of the addon. I'll send you the details via e-mail.

UV-mapping, textured buildings, terrain with satellite imagery will be in the premium version to be released within weeks.

**Comment by JeremyBYU** [comment]
I'll try to see if I can work with @vvoovv to figure out a way that we can turn it into an addon to the osm-blender plugin.  Here is the main class that handles the projections @cardoff. Email me if you have any other questions, I think we are off topic the licensing question now.

```python
"""
util/all_projections.py
This file is part of blender-osm (OpenStreetMap importer for Blender).
Copyright (C) 2014-2017 Vladimir Elistratov, Jeremy Castagno
prokitektura+support@gmail.com

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import math
from pyproj import Proj

DEFAULT_SRID = 'epsg:3857'

class AllProjections:
    radius = 6378137.

    def __init__(self, srid=DEFAULT_SRID, **kwargs):
        # setting default values
        self.lat = 0. # in degrees
        self.lon = 0. # in degrees

        # override default lat long coordinates
        for attr in kwargs:
            setattr(self, attr, kwargs[attr])
        # Set up user defined projection
        try:
            self.proj = Proj(init=srid)
        except:
            print('SRID is invalid! Defaulting to ' +  DEFAULT_SRID)
            self.proj = Proj(init=DEFAULT_SRID)

        self.offset = self.proj(self.lon, self.lat)

    def fromGeographic(self, lat, lon):
        proj_coords = self.proj(lon, lat)
        return (proj_coords[0] - self.offset[0], proj_coords[1] - self.offset[1], 0.)

    def toGeographic(self, x, y):
        lonlat_coords = self.proj(x + self.offset[0], y + self.offset[1], inverse=True)
        return (lonlat_coords[1], lonlat_coords[0])

```

**Comment by vvoovv** [comment]
For those awaiting the custom projections. The development is in the separate repo [bpyproj](https://github.com/JeremyBYU/bpyproj).

**Comment by vvoovv** [comment]
Great news!
The base version of the addon is now available for free [here](https://gumroad.com/l/blender-osm).

**Comment by vvoovv** [comment]
Great news again!
The source code is back to github. The code for the versions released publicly is located in the [release](https://github.com/vvoovv/blender-osm/tree/release) branch.
